src/event_listeners.jl

```diff
@@ -0,0 +1,139 @@
+        println("done.")
+
+        GitHub.post_status(event, sha, GitHub.SUCCESS,
+                           auth = auth, context = "Benchmarker",
+                           description = "Benchmarks complete!")
+
+        return HttpCommon.Response(200)
+    end
+
+    # Start the server on port 8000
+    GitHub.run(listener, 8000)
+
+"""
+immutable EventListener
+    server::HttpServer.Server
+    function EventListener(handle;
+                           auth::Authorization = AnonymousAuth(),
+                           secret = nothing,
+                           events = nothing,
+                           repos = nothing,
+                           forwards = nothing)
+        if !(isa(forwards, Void))
+            forwards = map(HttpCommon.URI, forwards)
+        end
+
+        server = HttpServer.Server() do request, response
+            try
+                if !(isa(secret, Void)) && !(is_valid_secret(request, secret))
+                    return HttpCommon.Response(400, "invalid signature")
+                end
+
+                if !(isa(events, Void)) && !(is_valid_event(request, events))
+                    return HttpCommon.Response(400, "invalid event")
+                end
+
+                payload = Requests.json(request)
+
+                if !(isa(repos, Void)) && !(is_valid_repo(payload, repos))
+                    return HttpCommon.Response(400, "invalid repo")
+                end
+
+                if !(isa(forwards, Void))
+                    for address in forwards
+                        Requests.post(address,
+                                      UTF8String(request.data),
+                                      headers=request.headers)
+                    end
+                end
+
+                event = Event(request, payload)
+
+                return handle(event, auth)
+            catch err
+                println("SERVER ERROR: $err")
+                return HttpCommon.Response(500)
+            end
+        end
+
+        server.http.events["listen"] = port -> begin
+            println("Listening for GitHub events sent to $port;")
+            println("Whitelisted events: $(isa(events, Void) ? "All" : events)")
+            println("Whitelisted repos: $(isa(repos, Void) ? "All" : repos)")
+        end
+
+        return new(server)
+    end
+end
+
+function Base.run(listener::EventListener, args...; kwargs...)
+    return HttpServer.run(listener.server, args...; kwargs...)
+end
+
+########################
+# Validation Functions #
+########################
+
+has_sig_header(request::HttpCommon.Request) = haskey(request.headers, "X-Hub-Signature")
+sig_header(request::HttpCommon.Request) = request.headers["X-Hub-Signature"]
+
+function is_valid_secret(request::HttpCommon.Request, secret::AbstractString)
+    if has_sig_header(request)
+        payload_string = UTF8String(request.data)
+        secret_sha = "sha1="*Nettle.hexdigest("sha1", secret, payload_string)
```

Comment:
A safe way to do this is to tag a new minor version as the first to be .4
only. The 1.0.* series would continue to be .3 compatible forever. You
could continue to push to that branch (eg, bug fixes), which .3 users would
automatically update from.

Meanwhile new development would take place in 1.1.*
On Fri, Nov 6, 2015 at 9:50 AM Jarrett Revels <notifications@github.com>
wrote:

> In src/event_listeners.jl
> <https://github.com/JuliaWeb/GitHub.jl/pull/39#discussion_r44151141>:
>
> > +
> > +function Base.run(listener::EventListener, args...; kwargs...)
> > +    return HttpServer.run(listener.server, args...; kwargs...)
> > +end
> > +
> > +########################
> > +# Validation Functions #
> > +########################
> > +
> > +has_sig_header(request::HttpCommon.Request) = haskey(request.headers, "X-Hub-Signature")
> > +sig_header(request::HttpCommon.Request) = request.headers["X-Hub-Signature"]
> > +
> > +function is_valid_secret(request::HttpCommon.Request, secret::AbstractString)
> > +    if has_sig_header(request)
> > +        payload_string = UTF8String(request.data)
> > +        secret_sha = "sha1="*Nettle.hexdigest("sha1", secret, payload_string)
>
> Adding this would mean updating GitHub.jl's REQUIRE to v0.4, yeah? I think
> my use of UTF8String above might also be an instance of
> constructor-fallback-to-convert, which is v0.4 behavior (could easily just
> do an explicit convert here to make it work v0.3, though).
>
> @WestleyArgentum <https://github.com/WestleyArgentum> are you okay with
> making the leap to v0.4 here? I know you expressed some concern in the
> other thread, but I think that trying to maintain backwards compatibility
> with v0.3 is only going to hinder us moving forward. People can still use
> v1.0.2 of GitHub.jl if they need v0.3 support.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaWeb/GitHub.jl/pull/39/files#r44151141>.
>